crates/core/src/client/mod.rs

```diff
@@ -499,12 +488,68 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
     }
 
     /// Returns the estimated gas for a transaction
-    async fn estimate_gas(
-        &self,
-        _call_request: CallRequest,
-        _block_number: Option<BlockId>,
-    ) -> Result<U256, EthApiError<P::Error>> {
-        todo!();
+    async fn estimate_gas(&self, request: CallRequest, block_id: BlockId) -> Result<U256, EthApiError<P::Error>> {
+        let chain_id: u64 = request
+            .chain_id
+            .ok_or_else(|| EthApiError::MissingParameterError("chain_id for estimate_gas".into()))?
+            .low_u64();
+
+        let from = request.from.ok_or_else(|| EthApiError::MissingParameterError("from for estimate_gas".into()))?;
+        let nonce: Felt252Wrapper = self.nonce(from, block_id).await?.try_into()?;
+
+        let gas_limit: Felt252Wrapper = request.gas.unwrap_or_else(|| U256::from(1_000_000)).try_into()?;
+        let max_fee_per_gas: Felt252Wrapper =
+            request.max_fee_per_gas.unwrap_or_else(|| U256::from(BASE_FEE_PER_GAS)).try_into()?;
+        let max_priority_fee_per_gas: Felt252Wrapper =
+            request.max_priority_fee_per_gas.unwrap_or_else(|| U256::from(MAX_PRIORITY_FEE_PER_GAS)).try_into()?;
+
+        let to = request.to.map_or(TransactionKind::Create, TransactionKind::Call);
+
+        let value: Felt252Wrapper = request.value.unwrap_or_else(|| U256::from(0)).try_into()?;
+
+        let data = request.data.unwrap_or_default();
+
+        let tx = Transaction::Eip1559(TxEip1559 {
```

Comment:
why do we convert some values to Felt252Wrapper before creating an Eth typed tx? 
for instance gas_limit 
or nonce

Reply:
Not much since indeed, I was trying to avoid having to map_err but I changed this and added a UintConversionError

Reply:
I don't really understand why we need to unbundle the `request`, to recreate a `tx`, since `request`

```
pub struct CallRequest {
    /// From
    pub from: Option<Address>,
    /// To
    pub to: Option<Address>,
    /// Gas Price
    pub gas_price: Option<U256>,
    /// EIP-1559 Max base fee the caller is willing to pay
    pub max_fee_per_gas: Option<U256>,
    /// EIP-1559 Priority fee the caller is paying to the block author
    pub max_priority_fee_per_gas: Option<U256>,
    /// Gas
    pub gas: Option<U256>,
    /// Value
    pub value: Option<U256>,
    /// Transaction data
    ///
    /// This accepts both `input` and `data`
    #[serde(alias = "input")]
    pub data: Option<Bytes>,
    /// Nonce
    pub nonce: Option<U256>,
    /// chain id
    pub chain_id: Option<U64>,
    /// AccessList
    pub access_list: Option<AccessList>,
    /// EIP-2718 type
    #[serde(rename = "type")]
    pub transaction_type: Option<U8>,
}
```

which already looks like a tx to me

Reply:
This is because I need to rlp encode the data I receive, easiest way I found was to unbundle the call request in a Transaction, which implements `encode_with_signature`.
See line `tx.encode_with_signature(&Signature::default(), &mut data, false);`